sessiontxn/isolation/repeatable_read.go

```diff
@@ -0,0 +1,265 @@
+	txnCxt.SetForUpdateTS(ts)
+	txn.SetOption(kv.SnapshotTS, ts)
+
+	p.forUpdateTS = ts
+
+	return
+}
+
+// updateForUpdateTS acquires the latest TSO and update the TransactionContext and kv.Transaction with it.
+func (p *PessimisticRRTxnContextProvider) updateForUpdateTS() (err error) {
+	sctx := p.sctx
+	var txn kv.Transaction
+
+	if txn, err = sctx.Txn(false); err != nil {
+		return err
+	}
+
+	if !txn.Valid() {
+		return errors.Trace(kv.ErrInvalidTxn)
+	}
+
+	// Because the ForUpdateTS is used for the snapshot for reading data in DML.
+	// We can avoid allocating a global TSO here to speed it up by using the local TSO.
+	version, err := sctx.GetStore().CurrentVersion(sctx.GetSessionVars().TxnCtx.TxnScope)
+	if err != nil {
+		return err
+	}
+
+	sctx.GetSessionVars().TxnCtx.SetForUpdateTS(version.Ver)
+	txn.SetOption(kv.SnapshotTS, sctx.GetSessionVars().TxnCtx.GetForUpdateTS())
+
+	return nil
+}
+
+// OnStmtStart is the hook that should be called when a new statement started
+func (p *PessimisticRRTxnContextProvider) OnStmtStart(ctx context.Context) error {
+	if err := p.baseTxnContextProvider.OnStmtStart(ctx); err != nil {
+		return err
+	}
+
+	p.forUpdateTS = 0
+	p.followingStmtIsPointGetForUpdate = false
+
+	return nil
+}
+
+// OnStmtRetry is the hook that should be called when a statement is retried internally.
+func (p *PessimisticRRTxnContextProvider) OnStmtRetry(ctx context.Context) (err error) {
+	if err = p.baseTxnContextProvider.OnStmtRetry(ctx); err != nil {
+		return err
+	}
+
+	txnCtxForUpdateTS := p.sctx.GetSessionVars().TxnCtx.GetForUpdateTS()
+	// If TxnCtx.forUpdateTS is updated in OnStmtErrorForNextAction, we assign the value to the provider
+	if txnCtxForUpdateTS > p.forUpdateTS {
+		p.forUpdateTS = txnCtxForUpdateTS
+	} else {
+		p.forUpdateTS = 0
+	}
+
+	p.followingStmtIsPointGetForUpdate = false
+
+	return nil
+}
+
+// OnStmtErrorForNextAction is the hook that should be called when a new statement get an error
+func (p *PessimisticRRTxnContextProvider) OnStmtErrorForNextAction(point sessiontxn.StmtErrorHandlePoint, err error) (sessiontxn.StmtErrorAction, error) {
+	switch point {
+	case sessiontxn.StmtErrAfterPessimisticLock:
+		return p.handleAfterPessimisticLockError(err)
+	default:
+		return sessiontxn.NoIdea()
+	}
+}
+
+// Advise is used to give advice to provider
+func (p *PessimisticRRTxnContextProvider) Advise(tp sessiontxn.AdviceType, val []any) error {
+	switch tp {
+	case sessiontxn.AdviceWarmUp:
+		return p.warmUp()
+	case sessiontxn.AdviceOptimizeWithPlan:
+		return p.optimizeWithPlan(val)
+	default:
+		return p.baseTxnContextProvider.Advise(tp, val)
+	}
+}
+
+// optimizeWithPlan optimizes for-update-like point get execution. We set p.followingStmtIsPointGetForUpdate
+// to be true in these cases. And when p.getForUpdateTs is called, we do not acquire tso from PD immediately but return the value of
+// txnCtx.GetForUpdateTS() expecting that the values that the point get operation acquires have not been changed.
```

Comment:
The description is a bit difficult to understand, you could describe it like this:
```
user cases/constraints: xxx
benefits: xxx
```

Reply:
I have fixed the comments.